Automattic/pocket-casts-android
# Sending refresh token from phone to watch

## Description
This PR automatically authenticates a user on a paired watch when they are logged into Pocket Casts on their phone by:

1. Retrieving a refresh token on the phone.
2. Sending that refresh token to the watch using the [Wearable Data Layer API](https://developer.android.com/training/wearables/data/data-layer).
3. When the watch receives a refresh token, the watch will use it to log in with that token.

One issue with the current implementation is that it's pretty pushy about logging you in on the watch if you're logged in on a connected phone. For example, if you're logged in on both devices, and you log out on the watch, then the next time you reopen the app you'll be logged back in on the watch. I'm not sure how I feel about this, it doesn't seem ideal, but it feels like an edge case for a user to want to be logged in on their phone but not their watch. If a user did want to keep them separate, they would just need to log into a different account on the watch. To me the current implementation feels good enough for now, but let me know if you have any thoughts.

https://user-images.githubusercontent.com/4656348/233832495-a3345807-9182-4fba-9239-ae877a930f13.mov

## Testing Instructions

When doing these tests, make sure you're using the same build variant on both the watch and the phone. If you don't, the syncing won't work. You may also want to comment out [this line](https://github.com/Automattic/pocket-casts-android/blob/d486b0ac7924402685a5f38422ce4d1796a29b8e/wear/src/main/kotlin/au/com/shiftyjelly/pocketcasts/wear/MainActivity.kt#L242) while you are testing so the notification doesn't get automatically dismissed too quickly. Note that I do have it set up so that the notification should show for a minimum of 5 seconds to ensure the user has time to see it and realize what is happening since this isn't a user initiated action but happens automatically when a refresh token is found.

1. Use a phone and a watch that are paired
2. Remove Pocket Casts from both the phone and the watch
3. Install the Pocket Casts app module from this branch onto the phone
5. Log into Pocket Casts on the phone
6. Install the Pocket Casts wear module from this branch onto the watch
7. ✅ Observe that the watch app is logged in pretty much immediately with a "logging in" screen shown while the refresh is in progress. This screen will show a spinning placeholder image if a gravatar image is not available for this account or if the gravatar image is still loading. I have this screen set to show for at least 5 seconds even if the refresh is finished earlier in order to make sure that there is enough time for the user to see this screen and realize that they are being logged in automatically.
8. ✅  After the refresh is complete, the "logging in" screen should be dismissed and the Now Playing and Up Next chips should be updated appropriately (assuming the account has a currently playing episode and it's Up Next queue is not empty)
9. Log out of the app on the phone
10. ✅  Observe that the watch app is still logged in
11. Log out of the watch app
12. Log into Pocket Casts on the phone
13. ✅ Observe that the watch app is logged in pretty much immediately

## Checklist
